### PR TITLE
Add docker log options and enable live restore

### LINKED ISSUE #1220: Docker daemon restart, recreates the containers
**Description**

If I ssh into the node and restart a docker daemon (`systemctl restart docker`) it restarts all the containers on the node.

**Impact**

Downtime to the workloads.

**Environment and steps to reproduce**

Any Lokomotive environment running on top of Flatcar.

**Expected behavior**

Daemon should restart without container downtime.

**Additional information**

To use the live restore feature add this config to `/etc/docker/daemon.json` using ignition:

```json
{
  "live-restore": true
}
```

And if you have made only config change and want docker daemon to load it, then just running `systemctl reload docker` is enough.

### LINKED ISSUE #1221: Configure default log storage size for docker daemon
**Current situation**

Right now we don't have default value for the log storage size of the containers.

**Impact**

There could be cases where a noisy application creates disk pressure on the node. Thereby either getting evicted or recreated. And depending on the criticality of the application it can incur downtime.

**Ideal future situation**

Either configure a sane value or allow user to choose what size they want across the machines in a worker pool.

**Implementation options**

Add following to `/etc/docker/daemon.json` via ignition:

```json
{
"log-driver": "json-file",
"log-opts": {
    "max-size": "10m",
    "max-file": "3"
    }
}
```

- `max-size`: Size of the log file here it is 10 megabytes.
- `max-file`: Number of log files to keep around of above size.

### PR DESCRIPTION
**Pull request by surajssd**
# Add docker log options and enable live restore

Earlier if the docker daemon was restarted the containers would be restarted as well. This PR enables live restore where containers are not restarted.

This PR also adds log options where if the container log size is more than 100MB then docker daemon creates new log file and only three such log files are kept around.

Fixes #1220
Fixes #1221 

# How to use

- Install a cluster on Packet or AWS or Baremetal with this change. To test the change also install these components to perform tests `openebs-operator`, `openebs-storage-class`, `prometheus-operator`, `contour`, `metallb`, `httpbin`, `external-dns`.
- ssh into any machine of the cluster and look for file `/etc/docker/daemon.json`.

# Testing done

- Start a pod in the cluster to stress the application from:

```bash
kubectl run -it --rm debug-network-$RANDOM --image-pull-policy=Always --image=quay.io/surajd/fedora-networking --restart=Never bash
```

- Install `httperf`:

```
dnf -y install httperf
```

- Run the perf:

```
httperf --hog --server <httpbin endpoint> --uri "/" --num-conn 1200000 --num-call 2 --timeout 5 --rate 4000 --port 80
```

Run this multiple times to increase the log size.

- Simultaenously goto the node and find the envoy directoy:

```
cd /var/lib/docker/containers/$(docker inspect --format='{{.Id}}' $(docker ps | grep envoyproxy | awk '{print $1}'))
watch du -sh ./*
```

Keep an eye on the `*.log*` files. The size of each `*.log*` won't go beyond 100MB, and new files will be created ending with `.log.1` and `.log.2`.

- Optional: Watch the envoy logs

```
kubectl logs -n projectcontour -f -l app=envoy
```

### CODE REVIEW

**Comment by surajssd** [comment]
Suggestions for testing this are welcome.

**Comment by surajssd** [comment]
**NOTE**: Applying live restore on a non-live restore node will restart all the containers on it. It might take couple of minutes before they start. This does not mean that if a user runs `lokoctl cluster apply` against a lokomotive cluster it will restart everything. Right now the node recreation on ignition change is disabled. But any new worker pool will come with live restore enabled.

**Comment by surajssd** [comment]
## Manual update steps:

If a user wants to enable this manually in a cluster on any node then follow the following steps:

- SSH into the node and become root with `sudo -s`.
- Create the docker config file:

```json
echo '
{
  "live-restore": true,
  "log-opts": {
    "max-size": "100m",
    "max-file": "3"
  }
}
' | tee /etc/docker/daemon.json
```

- Restart docker daemon:

**NOTE**: This will restart all the containers on the node.

```
systemctl restart docker
```

File: assets/terraform-modules/aws/flatcar-linux/kubernetes/cl/controller.yaml.tmpl
Comment (invidian) [inline assets/terraform-modules/aws/flatcar-linux/kubernetes/cl/controller.yaml.tmpl:275]: Why 100m and 3 (and not 200m and 2)?

We could also have some note in the documentation, that logs will be limited to last 300MB for each container.

**Review by invidian — CHANGES_REQUESTED** [review]
About testing, I think we could have a DaemonSet which has access to Docker socket and runs `docker info -f {{.LiveRestoreEnabled}}` etc to make sure that those features are effective.

Otherwise looks good.

File: assets/terraform-modules/aws/flatcar-linux/kubernetes/cl/controller.yaml.tmpl
Comment (surajssd) [reply assets/terraform-modules/aws/flatcar-linux/kubernetes/cl/controller.yaml.tmpl:275]: No specific reason to keep it to `100m x 3` vs `200m x 2`. A default to start with, is all.

Comment (invidian) [reply assets/terraform-modules/aws/flatcar-linux/kubernetes/cl/controller.yaml.tmpl:275]: I think we could make some assumptions here. Let's say minimal disk size of 20GB and max number of pods per node on 2GB or 4GB machine and assume they all log extensively.

Comment (surajssd) [reply assets/terraform-modules/aws/flatcar-linux/kubernetes/cl/controller.yaml.tmpl:275]: Assuming that 20GB is exclusively available to store logs:

- With 300MB limit we can fit 66 containers,
- With 400MB limit we can fit 50 containers.

Comment (invidian) [reply assets/terraform-modules/aws/flatcar-linux/kubernetes/cl/controller.yaml.tmpl:275]: Hm, default for `--max-pods` is 110. I wonder how that correlates.

File: test/system/docker_live_restore_test.go
Comment (knrt10) [inline test/system/docker_live_restore_test.go:54]: I don't mind this, but can we push this image to our quay, so that we don't have to come to fix this issue again in the future. What do you think @surajssd ?

File: test/system/docker_live_restore_test.go
Comment (invidian) [inline test/system/docker_live_restore_test.go:60]: I think using just `grep` for the output makes test much simpler, as it can be then just tested with `kubectl`.

File: docs/configuration-reference/platforms/aws.md
Comment (invidian) [inline docs/configuration-reference/platforms/aws.md:271]: Maybe we can put this in `docs/concepts/logging.md`? Then we can deduplicate and provide more information like what are our assumptions here.

**Review by invidian — CHANGES_REQUESTED** [review]
As Kubernetes 1.20 is deprecating Docker as container runtime, I would like to know if there is an equivalent solution for log rotation for containerd, so we know how easy it will be to migrate to it.

I would like to avoid adding a solution now, which will have to be replaced very soon anyway.

File: test/system/docker_live_restore_test.go
Comment (surajssd) [reply test/system/docker_live_restore_test.go:60]: I am not sure I understand what you mean by using `grep` and `kubectl`?

**Comment by surajssd** [comment]
> As Kubernetes 1.20 is deprecating Docker as container runtime, I would like to know if there is an equivalent solution for log rotation for containerd, so we know how easy it will be to migrate to it.

This is not finalised yet. It is still a topic of discussion.

**Comment by invidian** [comment]
>This is not finalised yet. It is still a topic of discussion.

We should switch to containerd anyway, as it is more lightweight than Docker.

File: test/system/docker_live_restore_test.go
Comment (invidian) [reply test/system/docker_live_restore_test.go:60]: `docker info -f {{.LiveRestoreEnabled}} | grep true && sleep infinity` and you don't need to read logs. You can also copy-paste manifest to YAML file, do `kubectl apply` and manually see if it converges.

File: docs/concepts/logging.md
Comment (invidian) [inline docs/concepts/logging.md:8]: Please justify the numbers, associated assumptions and explain the limitations of this configuration.

**Review by invidian — CHANGES_REQUESTED** [review]

File: docs/concepts/logging.md
Comment (surajssd) [reply docs/concepts/logging.md:8]: > Lokomotive restricts the logs per container to be 300MB divided in three files (100MB per file). Due to this limitation any log intensive app will have only last 300MB of logs anything beyond that will be purged by Docker. To preserve logs of such applications for long term access use a log shipping solution.

How about ^ this?

File: test/system/docker_live_restore_test.go
Comment (surajssd) [reply test/system/docker_live_restore_test.go:60]: This works. But the problem still remains for the test code to know if the test has passed. What will be the exit condition? In current case we look at logs and match with expected output and tests exit.

But in this case should we check if the pods are running for a while? What should the exit condition be?

Comment (invidian) [reply test/system/docker_live_restore_test.go:60]: Just make it wait for daemonset to converge. if it fails, it will go into crash loopback and never converge.

File: docs/concepts/logging.md
Comment (invidian) [reply docs/concepts/logging.md:8]: I would add, that this should allow to run around 50 logging intensive containers on 20GB disk (Flatcar takes a bit of space, but we calculated limit of 66, so saying 50 sounds reasonable I think). And I would note, this is a Docker _containers_ and not pods, as one pod consist of at least 2 containers.

Comment (surajssd) [reply docs/concepts/logging.md:8]: > Lokomotive restricts the logs per _Docker container_ (not _Kubernetes pods_) to be 300MB divided into three files (100MB per file). This should allow running around fifty log intensive containers on a 20GB disk. Due to this limitation, any log intensive app will have only last 300MB of logs anything beyond that will be purged by Docker. To preserve logs of such applications for long term access use a log shipping solution.

Comment (invidian) [reply docs/concepts/logging.md:8]: Sounds great :+1: Thanks!

File: assets/terraform-modules/aws/flatcar-linux/kubernetes/cl/controller.yaml.tmpl
Comment (invidian) [reply assets/terraform-modules/aws/flatcar-linux/kubernetes/cl/controller.yaml.tmpl:275]: We documented the constraints, so marking as resolved.

File: test/system/docker_live_restore_test.go
Comment (invidian) [inline test/system/docker_live_restore_test.go:68]: We should also have e2e test for `log-opts`.

File: test/system/docker_live_restore_test.go
Comment (invidian) [inline test/system/docker_live_restore_test.go:71]: I suggest more descriptive test name:
```suggestion
func TestDockerHasLiveRestoreEnabledOnAllNodes(t *testing.T) {
```

File: test/system/docker_live_restore_test.go
Comment (invidian) [inline test/system/docker_live_restore_test.go:89]: A minute seems a lot to me. Is this really necessery? We run other tests using similar pattern and we never wait. Perhaps logic of `WaitForDaemonSet` is somehow incorrect then?

Comment (invidian) [reply test/system/docker_live_restore_test.go:89]: Perhaps using Init contianer to run the check and run sleep as main container would allow us to avoid waiting?

**Review by invidian — CHANGES_REQUESTED** [review]

File: test/system/docker_live_restore_test.go
Comment (surajssd) [reply test/system/docker_live_restore_test.go:89]: One minute is fine, since the test is running in parallel. The logic of `WaitForDaemonset` is to verify if the pods come up alright, assuming that the workloads have correct readiness or liveness probes set.

Here we are abusing the function `WaitForDaemonset`, the test should have been a job where one is scheduled on each node. And then verify if all jobs passed. But for convenience we are deploying it as DS and then waiting for a bit so that it goes into CrashLoopBackOff.

Comment (invidian) [reply test/system/docker_live_restore_test.go:89]: I'd use init container and skip the wait. It seems hacky.

Comment (surajssd) [reply test/system/docker_live_restore_test.go:89]: I think reverting back to checking logs and matching with what is expected is the most reliable solution.

File: test/system/docker_live_restore_test.go
Comment (iaguis) [inline test/system/docker_live_restore_test.go:79]:
```suggestion
		t.Fatalf("Failed unmarshaling manifest: %v", err)
```

File: test/system/docker_live_restore_test.go
Comment (iaguis) [inline test/system/docker_live_restore_test.go:84]:
```suggestion
		t.Fatalf("Failed to create DaemonSet: %v", err)
```

File: test/system/docker_live_restore_test.go
Comment (iaguis) [inline test/system/docker_live_restore_test.go:97]:
```suggestion
			t.Logf("Failed to remove DaemonSet: %v", err)
```

File: test/system/docker_live_restore_test.go
Comment (iaguis) [reply test/system/docker_live_restore_test.go:89]: @surajssd I agree that 1 minute is fine, but I think it's better to avoid having random sleeps. What's the issue with using an init container to do the test? In that case the pod won't be ready until the init container exits successfully and we can just use WaitForDaemonSet.

File: test/system/docker_live_restore_test.go
Comment (iaguis) [reply test/system/docker_live_restore_test.go:68]: Yes, this should be similar.

File: assets/terraform-modules/aws/flatcar-linux/kubernetes/cl/controller.yaml.tmpl
Comment (iaguis) [inline assets/terraform-modules/aws/flatcar-linux/kubernetes/cl/controller.yaml.tmpl:271]: I was reading about this in the [Docker documentation](https://docs.docker.com/config/containers/live-restore/) and it seems to not be supported for upgrading Major releases. However, AFAIU a Docker upgrade would happen by rebooting, so I guess we're fine here. Just wanted to make that explicit with this comment, no action needed :).

**Review by iaguis — CHANGES_REQUESTED** [review]

File: test/system/docker_live_restore_test.go
Comment (invidian) [inline test/system/docker_live_restore_test.go:124]: I guess this info is not available from `docker info`?

File: test/system/docker_live_restore_test.go
Comment (invidian) [inline test/system/docker_live_restore_test.go:141]: Reported test names will be `TestDockerNodeConfig/Docker_Live_Restore`. Can we change keys sub-test names to `has_live_restore_enabled` etc? So then test looks like: `TestDockerNodeConfig/has_live_restore_enabled`.

File: test/system/docker_live_restore_test.go
Comment (invidian) [inline test/system/docker_live_restore_test.go:138]: I think `map[string]string` will be simpler here.

File: test/system/docker_live_restore_test.go
Comment (invidian) [inline test/system/docker_live_restore_test.go:150]: Client can be shared between tests.

Comment (invidian) [reply test/system/docker_live_restore_test.go:150]: Maybe change it to `dsClient := client.AppsV1().DaemonSets("kube-system")`, so we can avoid variable for `kube-system` and shorten lines, which improves readability.

File: test/system/docker_live_restore_test.go
Comment (invidian) [inline test/system/docker_live_restore_test.go:118]: How is this image maintained? Maybe we can bind-mount `jq` from host?

**Review by invidian — CHANGES_REQUESTED** [review]
Some technical nits.

File: test/system/docker_live_restore_test.go
Comment (surajssd) [reply test/system/docker_live_restore_test.go:150]: `namespace` var is used in one more place. I will keep that var as it is.

File: test/system/docker_live_restore_test.go
Comment (surajssd) [reply test/system/docker_live_restore_test.go:138]: Yep it would be concise, but this is explicit on what each field is.

File: test/system/docker_live_restore_test.go
Comment (surajssd) [reply test/system/docker_live_restore_test.go:124]: Yep this wasn't exposed via `docker info`.

File: test/system/docker_live_restore_test.go
Comment (invidian) [reply test/system/docker_live_restore_test.go:138]: Fair enough.

File: test/system/docker_live_restore_test.go
Comment (invidian) [reply test/system/docker_live_restore_test.go:118]: @surajssd what about this?

File: test/system/docker_live_restore_test.go
Comment (invidian) [inline test/system/docker_live_restore_test.go:139]:
```suggestion
		{"has_live_restore_enabled", dockerLiveRestoreOnNodesDSManifest},
```

File: test/system/docker_live_restore_test.go
Comment (invidian) [inline test/system/docker_live_restore_test.go:140]:
```suggestion
		{"has_log_options_set", dockerLogOptsOnNodesDSManifest},
```

File: test/system/docker_live_restore_test.go
Comment (invidian) [reply test/system/docker_live_restore_test.go:139]: So we have nice test name describing the property of the system: `TestDockerNodeConfig/has_live_restore_enabled`.

Also using `has_live_restore_enabled` over `Has Live Restore Enabled` make test _greppable_, as mentioned earlier, as `Has Live Restore Enabled` would be converted to `Has_Live_Restore_Enabled` anyway, which makes it harder to find.

File: test/system/docker_live_restore_test.go
Comment (invidian) [inline test/system/docker_live_restore_test.go:163]: Let's use standard timeout and retry interval.
```suggestion
			testutil.WaitForDaemonSet(t, client, namespace, ds.Name, testutil.RetryInterval, testutil.Timeout)
```

File: test/system/docker_live_restore_test.go
Comment (invidian) [inline test/system/docker_live_restore_test.go:167]:
```suggestion
					t.Logf("Failed to delete DaemonSet: %v", err)
```

**Review by invidian — DISMISSED** [review]
Nits, otherwise looks good.

**Review by iaguis — COMMENTED** [review]
lgtm after nits are fixed

File: test/system/docker_live_restore_test.go
Comment (surajssd) [reply test/system/docker_live_restore_test.go:118]: Can't use this because it errors as:

```
jq: error while loading shared libraries: libjq.so.1: cannot open shared object file: No such file or directory
```

File: test/system/docker_live_restore_test.go
Comment (surajssd) [reply test/system/docker_live_restore_test.go]: Added underscores, but the name `Is` prefixed sounds right than with `has`.

File: test/system/docker_live_restore_test.go
Comment (surajssd) [reply test/system/docker_live_restore_test.go]: Here we are checking the exact values and not merely if they exists. So `has` does not make sense or it does not convey the exact meaning.

File: test/system/docker_live_restore_test.go
Comment (knrt10) [inline test/system/docker_live_restore_test.go:138]: We could have the same convention as below.

```suggestion
		{"Docker_Live_Restore_Enabled", dockerLiveRestoreOnNodesDSManifest},
```

**Review by knrt10 — CHANGES_REQUESTED** [review]
Small NIT, rest LGTM.

File: test/system/docker_live_restore_test.go
Comment (invidian) [reply test/system/docker_live_restore_test.go:117]: And if we mount the library as well? :smile:

Comment (invidian) [reply test/system/docker_live_restore_test.go:117]: Ah, but it depends also on other libs... Still, how is the image built/maintained?

File: test/system/docker_live_restore_test.go
Comment (invidian) [inline test/system/docker_live_restore_test.go:138]: `TestDockerNodeConfig/Is_Live_Restore_Enabled` does not read right to me. Better would be `TestDockerNodeConfig/Live_Restore_Is_Enabled`. But I still think `TestDockerNodeConfig/Has_Live_Restore_Enabled` is the best name, as it's a declarative statement.

Comment (surajssd) [reply test/system/docker_live_restore_test.go:138]: My bad, will remove the docker from the name in the other test. The invoking test is already called `TestDockerNodeConfig` so not repeating the `Docker` in it.

File: test/system/docker_live_restore_test.go
Comment (surajssd) [reply test/system/docker_live_restore_test.go:117]: Right now I am building and pushing it. I haven't put the Dockerfile anywhere.

```dockerfile
FROM registry.fedoraproject.org/fedora:32

RUN dnf -y install --setopt=tsflags=nodocs jq && \
    dnf clean all
```

Comment (invidian) [reply test/system/docker_live_restore_test.go:117]: Maybe we can put it as a comment or in commit message, so it's easier to find in the future?

**Review by iaguis — APPROVED** [review]
lgtm

**Review by knrt10 — APPROVED** [review]
+1

**Review by invidian — COMMENTED** [review]
I still think we could improve test case names:
>TestDockerNodeConfig/Is_Live_Restore_Enabled does not read right to me. Better would be TestDockerNodeConfig/Live_Restore_Is_Enabled. But I still think TestDockerNodeConfig/Has_Live_Restore_Enabled is the best name, as it's a declarative statement.

**Comment by iaguis** [comment]
This needs rebase. Personally I don't mind the test name too much and I won't block this on the change but while you're at it you could change it as @invidian says.

**Comment by surajssd** [comment]
> `TestDockerNodeConfig/Is_Live_Restore_Enabled` does not read right to me. Better would be `TestDockerNodeConfig/Live_Restore_Is_Enabled`. But I still think `TestDockerNodeConfig/Has_Live_Restore_Enabled` is the best name, as it's a declarative statement.

IMHO, the test name makes sense in the question form, because we are checking that if that behaviour is working fine!

**Review by iaguis — APPROVED** [review]

**Review by knrt10 — APPROVED** [review]

**Pull request merged**
